verilog/tools/extractor/verilog_extractor.cc

```diff
@@ -0,0 +1,332 @@
+                                         absl::string_view)>;
+std::map<int, TagExtractor> tagExtractor;
+
+std::vector<Block> Extract(const verible::SyntaxTreeLeaf &node, absl::string_view base) {
+    std::cout << "Start Leaf" << std::endl;
+    std::cout << verilog::NodeEnumToString(static_cast<verilog::NodeEnum>(node.Tag().tag)) << " <<>> " << node.Tag().tag
+              << " " << node.get()
+              << std::endl;
+    std::cout << "End Leaf" << std::endl;
+    std::cout << std::endl;
+    std::vector<Block> blocks;
+    return blocks;
+}
+
+std::vector<Block> Extract(const verible::SyntaxTreeNode &node, absl::string_view base) {
+    std::cout << "Start Node" << std::endl;
+    std::cout << verilog::NodeEnumToString(static_cast<verilog::NodeEnum>(node.Tag().tag)) << "  ";
+    std::cout << node.children().size() << std::endl;
+
+    std::vector<Block> blocks;
+    auto resolver = tagExtractor.find(node.Tag().tag);
+    if (resolver != tagExtractor.end()) {
+        blocks.push_back(resolver->second(node, node, base));
+        return blocks;
+    }
+
+    for (const auto &child : node.children()) {
+        if (child) {
+            if (child->Kind() == verible::SymbolKind::kNode) {
+                auto innerBlock = Extract(verible::SymbolCastToNode(*child), base);
+                blocks.insert(blocks.end(), innerBlock.begin(), innerBlock.end());
+            } else {
+                auto innerBlock = Extract(verible::SymbolCastToLeaf(*child), base);
+                blocks.insert(blocks.end(), innerBlock.begin(), innerBlock.end());
+            }
+        }
+    }
+
+    std::cout << "End Node" << std::endl << std::endl;
+    return blocks;
+}
+
+const verible::Symbol *GetChildByTag(const verible::SyntaxTreeNode &root, verilog::NodeEnum tag) {
+    for (const auto &child : root.children()) {
+        if (child) {
+            if (child->Tag().tag == static_cast<int>(tag)) {
+                return child.get();
+            }
+        }
+    }
+    return nullptr;
+}
+
+const verible::Symbol *GetFirstChildByTag(const verible::SyntaxTreeNode &root, verilog::NodeEnum tag) {
+    auto target = GetChildByTag(root, tag);
+    if (target == nullptr) {
+        for (const auto &child : root.children()) {
+            if (child && child->Kind() == verible::SymbolKind::kNode) {
+                target = GetFirstChildByTag(verible::SymbolCastToNode(*child), tag);
+                if (target != nullptr) {
+                    return target;
+                }
+            }
+        }
+    }
+    return target;
+}
+
+Block ExtractModuleInstantiation(const verible::SyntaxTreeNode &node, const verible::SyntaxTreeNode &root,
+                                 absl::string_view base) {
+
+    auto instantiationBase = GetChildByTag(node, verilog::NodeEnum::kInstantiationBase);
+
+    auto instantiationType = GetChildByTag(verible::SymbolCastToNode(*instantiationBase),
```

Comment:
You didn't check if instantiationBase is nullptr before dereferencing it.

GetChildByTag & GetFirstChildByTag can be improved by forcing the client to check the result (e.g., by using absl::StatusOr (not yet available) or various optional<> types -- similar exists in many languages). I'm not saying we should change GetChildByTag & GetFirstChildByTag now. It's fine to keep them. But make sure to check if instantiationBase is nullptr or not (same below for the rest of variables)

Reply:
@fangism do we have something like absl::statusOr for checking this nullptr case?

Reply:
No `StatusOr`, but a lot of related checked-accessor functions can be found in:

https://github.com/google/verible/blob/master/common/text/tree_utils.h